kani-compiler/src/kani_compiler.rs

```diff
@@ -51,16 +53,53 @@ pub fn run(args: Vec<String>) -> ExitCode {
     }
 }
 
-/// Configure the cprover backend that generate goto-programs.
-#[cfg(feature = "cprover")]
+/// Configure the boogie backend that generates boogie programs.
+fn boogie_backend(_queries: Arc<Mutex<QueryDb>>) -> Box<dyn CodegenBackend> {
+    #[cfg(feature = "boogie")]
+    return Box::new(BoogieCodegenBackend::new(_queries));
+    #[cfg(not(feature = "boogie"))]
+    rustc_session::early_error(
+        ErrorOutputType::default(),
+        "`--backend boogie` requires enabling the `boogie` feature",
+    );
+}
+
+/// Configure the cprover backend that generates goto-programs.
+fn cprover_backend(_queries: Arc<Mutex<QueryDb>>) -> Box<dyn CodegenBackend> {
+    #[cfg(feature = "cprover")]
+    return Box::new(GotocCodegenBackend::new(_queries));
+    #[cfg(not(feature = "cprover"))]
+    rustc_session::early_error(
+        ErrorOutputType::default(),
+        "`--backend cprover` requires enabling the `cprover` feature",
+    );
+}
+
+#[cfg(any(feature = "cprover", feature = "boogie"))]
 fn backend(queries: Arc<Mutex<QueryDb>>) -> Box<dyn CodegenBackend> {
-    Box::new(GotocCodegenBackend::new(queries))
+    let backend = queries.lock().unwrap().backend;
```

Comment:
Just be aware that this logic is rather fragile and we might want to improve this in the near future. There is no guarantee on the order that rustc build the backend object and call `Callbacks::config()`. If the order is inverted, the queries may not be initialized. That is the reason why we share the same QueryDb instead of copying the object at this level.